### ISSUE TITLE
The `Save` Button for notebook is not accessible in JLab 0.35.0

### ISSUE THREAD

**Issue by dclong** [jupyterlab/jupyterlab#5454]
**Describe the bug**
The `Save` button is gray after a notebook is changed. The notebook can still be saved by shortcut or from the `File` menu.

![image](https://user-images.githubusercontent.com/824507/46592614-27e74180-caf7-11e8-8267-cd20f815e7b8.png)

**To Reproduce**

1. Open a notebook in JupyterLab and make some changes
2. You will see that the `Save` button is gray and you cannot use it to save the notebook. 

**Expected behavior**
The user should be able to use the `Save` button to save the notebook when it's changed. 

**Desktop (please complete the following information):**

- OS: Ubuntu 18.04 via Docker image (in which JupyterLab is installed)
- Browser: firefox 62.0
- JupyterLab: 0.35.0

<details><summary>Troubleshoot Output</summary>
<pre>
Paste the output from running `jupyter troubleshoot` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.
</pre>
</details>

<details><summary>Browser Output</summary>
<pre>
Paste the output from your browser console here.
</pre>
</details>

**Comment by dharmaquark** [comment]
The change from that PR was to set the toolbar save sensitivity to:
        panel &&
        panel.context &&
        panel.context.contentsModel &&
        panel.context.contentsModel.writable

Could that be problematic for different browsers? Or are there cases where writable is undefined? I can put in a patch to use a more conservative check (only disable when writable is defined and explicitly ‘false’?)

**Comment by ian-r-rose** [comment]
I think that is only evaluated once, when the toolbar is created. We may need to change the new React-based toolbar API to take a function-or-boolean for the `enabled` attribute, rather than a boolean, as it is currently.

**Comment by jasongrout** [comment]
Good point. We should have caught this in review.  Sorry for not catching this, @dharmaquark!

@ian-r-rose - sounds like a great plan: +1.

**Comment by jasongrout** [comment]
Actually, digging deeper: the problem seems to be that the toolbar is created *before* the contentsModel is available (i.e., it is null), so the `panel.context.contentsModel` part of the test fails.

**Comment by dharmaquark** [comment]
ah! is there another way to check writable before the contentsModel is available? or would that involve a wider set of changes? if we move to a function-or-boolean, would it be possible for the sensitivity to update dynamically once the contentsModel is available/once the writable property is non-null?

**Comment by jasongrout** [comment]
I think it does a wider set of changes - something to trigger a rerender anytime the contentsModel becomes available, and perhaps also if the writable attribute changes.

**Comment by dharmaquark** [comment]
got it- I can put in a patch to revert for now in the short term if that works. Would be nice I think, if possible, to set the sensitivity of the different buttons dynamically in a way that doesn't cause a full rerender

**Comment by dharmaquark** [comment]
a related question- I noticed that in creating the toolbar button for save, we also check panel.context.model.readOnly, would that be something reasonable to use for now? (in theory, would that model.readOnly always be in sync with contentsModel.writable?)

**Comment by dharmaquark** [comment]
might make sense to check that for sensitivity, since we're already checking it in onClick (maybe the toolbar updates to 'Document is read-only' when readOnly is true?)

**Comment by ian-r-rose** [comment]
I think that for now we should revert the `enabled` change for the notebook toolbar, since we want to get the fix in for `0.35.1`. Restructuring how enabled-state is handled in the toolbar is a potentially breaking change, and a larger task.

**Comment by ian-r-rose** [comment]
Ooh, that could be a good fix as well.  Possibly instead of #5475, if you think it would be easy enough to do.

**Comment by jasongrout** [comment]
I've also been diving into this and came to the same conclusion, that we should have more command buttons. I found a couple of things:

1. We need to update things a bit more to finish off the toolbar changes that went into 0.35:

```patch
@@ -177,28 +184,6 @@ export namespace ToolbarItems {
     return new CellTypeSwitcher(panel.content);
   }
 
-  /**
-   * Add the default items to the panel toolbar.
-   */
-  export function populateDefaults(panel: NotebookPanel): void {
-    let toolbar = panel.toolbar;
-    toolbar.addItem('save', createSaveButton(panel));
-    toolbar.addItem('insert', createInsertButton(panel));
-    toolbar.addItem('cut', createCutButton(panel));
-    toolbar.addItem('copy', createCopyButton(panel));
-    toolbar.addItem('paste', createPasteButton(panel));
-    toolbar.addItem('run', createRunButton(panel));
-    toolbar.addItem('interrupt', Toolbar.createInterruptButton(panel.session));
-    toolbar.addItem('restart', Toolbar.createRestartButton(panel.session));
-    toolbar.addItem('cellType', createCellTypeItem(panel));
-    toolbar.addItem('spacer', Toolbar.createSpacerItem());
-    toolbar.addItem('kernelName', Toolbar.createKernelNameItem(panel.session));
-    toolbar.addItem(
-      'kernelStatus',
-      Toolbar.createKernelStatusItem(panel.session)
-    );
-  }
-
   /**
    * Get the default toolbar items for panel
    */
diff --git a/packages/notebook/src/widgetfactory.ts b/packages/notebook/src/widgetfactory.ts
index 5852b3781..199be2dcf 100644
--- a/packages/notebook/src/widgetfactory.ts
+++ b/packages/notebook/src/widgetfactory.ts
@@ -66,8 +66,7 @@ export class NotebookWidgetFactory extends ABCWidgetFactory<
    * Create a new widget.
    *
    * #### Notes
-   * The factory will start the appropriate kernel and populate
-   * the default toolbar items using `ToolbarItems.populateDefaults`.
+   * The factory will start the appropriate kernel.
    */
   protected createNewWidget(
     context: DocumentRegistry.IContext<INotebookModel>
```

2. The problem with creating a command toolbar button is that it needs access to the command registry, which is not available to the widget right now, I think. Also, the command id is set by the *extension*, so lives at a layer above the notebook widget itself. So perhaps the extension should populate (override?) the toolbar buttons with special command toolbar buttons, using our new method for overriding the default toolbar buttons?

**Comment by dharmaquark** [comment]
ah, @jasongrout, good point- I guess that takes us back to exploring whether there's a way to trigger a rerender anytime the contentsModel becomes available and/or when the writable attribute changes.
